### PR TITLE
noobaa/core: Transaction related changes

### PR DESCRIPTION
**Pull request by aspandey**
This patch makes changes to send bulk queries for
multiple collections in one transaction for postgres DB. In case of failure of any query, all the bulk operations will be rollbacked.

unit test has not been added for this patch

### CODE REVIEW

**Comment by aspandey** [comment]
@dannyzaken

File: src/server/system_services/system_store.js [hunk 1]
```diff
@@ -38,9 +38,11 @@ const size_utils = require('../../util/size_utils');
 const os_utils = require('../../util/os_utils');
 const config = require('../../../config');
 const db_client = require('../../util/db_client');
+const pgc = require('../../util/postgres_client');
 
 const { RpcError } = require('../../rpc');
 const master_key_manager = require('./master_key_manager');
+const { ok } = require('assert');
```
Comment (liranmauda) [inline src/server/system_services/system_store.js:45]: I think it is leftover, should we remove it?

File: src/server/system_services/system_store.js [hunk 2]
```diff
@@ -617,6 +619,8 @@ class SystemStore extends EventEmitter {
 
     async _make_changes_internal(changes) {
         const bulk_per_collection = {};
+        let bulk_results = [];
+        let pg_bulk_results = [];
```
Comment (liranmauda) [inline src/server/system_services/system_store.js:623]: Look like it is not beeing used.

File: src/util/postgres_client.js [hunk 3]
```diff
@@ -478,6 +480,79 @@ class BulkOp {
...
+        ok = true;
+    } catch (err) {
+        errmsg = err;
+        dbg.error('PgTransaction execute error', err);
+        if (should_rollback) await transaction.rollback();
+    } finally {
+        transaction.release();
+    }
+    return [errmsg, ok];
+    // return {
```
Comment (liranmauda) [inline src/util/postgres_client.js:525]: can we clean this comment, or do we want to keep it?

**Comment by liranmauda** [comment]
> Thanks @liranmauda for comments. This is work in progress. Will make these changes in next iteration.

Yes, noticed the draft after I reviewed 🙂

File: src/server/system_services/system_store.js [hunk 4]
```diff
@@ -740,17 +744,27 @@ class SystemStore extends EventEmitter {
...
 
-        const bulk_results = await Promise.all(Object.values(bulk_per_collection).map(
-            bulk => bulk.length && bulk.execute({ j: true })
-        ));
 
-        for (const res of bulk_results) {
-            if (res && !res.ok) {
-                dbg.error('got error on bulk execute', res.err);
-                throw new Error(res.err);
+        if (config.DB_TYPE === 'postgres') {
```
Comment (dannyzaken) [inline src/server/system_services/system_store.js:748]: Maybe worth hiding the different implementations in the db clients. I suggest adding a function to `DBClient` interface in [nb.d.ts](https://github.com/noobaa/noobaa-core/blob/aeb5e82334ba1526cec2016ad698045e44ba02dc/src/sdk/nb.d.ts#L649), and then you can have a different implementation for postgres\mongo. Maybe something like `execute_multiple_bulks()`.
This function should return a common `bulk_results` array that should be inspected the same for both DBs

File: src/util/postgres_client.js (hunk 3 above)
Comment (dannyzaken) [reply src/util/postgres_client.js:525]: in addition to my comment [here](https://github.com/noobaa/noobaa-core/pull/7249#discussion_r1155319425), we should return here the same structure that we return for mongo and handle it in a common code

File: src/server/system_services/system_store.js (hunk 4 above)
Comment (aspandey) [reply src/server/system_services/system_store.js]: Done.  Good Idea.

File: src/util/postgres_client.js (hunk 3 above)
Comment (aspandey) [reply src/util/postgres_client.js]: Done.

File: src/util/postgres_client.js [hunk 5]
```diff
@@ -478,6 +479,87 @@ class BulkOp {
...
 }
 
+async function multi_collection_bulk(bulk_per_collection) {
+
+    const client = PostgresClient.instance();
+
+    const queries = _.flatten(Object.values(bulk_per_collection).map(bulk => bulk.get_queries()));
+    const transaction = new PgTransaction(client);
+
+    let ok = false;
```
Comment (romayalon) [inline src/util/postgres_client.js:489]: this is the same code as execute() of bulkOp class, I don't see a value in maintaining this code in 2 different functions, easily can move to a helper function

File: src/server/system_services/system_store.js [hunk 6]
```diff
@@ -740,9 +740,8 @@ class SystemStore extends EventEmitter {
             });
         });
 
-        const bulk_results = await Promise.all(Object.values(bulk_per_collection).map(
-            bulk => bulk.length && bulk.execute({ j: true })
-        ));
+
+        const bulk_results = await db_client.instance().execute_multiple_bulks(bulk_per_collection);
```
Comment (romayalon) [inline src/server/system_services/system_store.js:744]: rollbacking all collections queries on failures of related collections in a single make_changes() can be a great value for our db validity, 
but it can cause unexpected failures if a single make_changes() updates/inserts/removes non related queries (rollback one query due to failure of non related query)
I would check that every call to make_changes() make sense in that way.

File: src/util/postgres_client.js (hunk 5 above)
Comment (aspandey) [reply src/util/postgres_client.js:372]: Makes sense. Done.

File: src/server/system_services/system_store.js (hunk 6 above)
Comment (aspandey) [reply src/server/system_services/system_store.js:750]: @romayalon,
I am not sure what could be not related queries in a single make_changes() function call.
On what basis we can say that two queries are related or not. This could be very complex to handle and could be bug prone if we do not have simple way of segregating queries. 
I have a limited understanding on this so could not visualize the tradeoff. 
@dannyzaken please provide your input as well.

For the refreshed patch, I have not made changes for this comment. Once we decide what to do next, I will send  modified code.

**Review by dannyzaken — COMMENTED** [review]
@aspandey, I discussed with @romayalon, and we agreed this change is ok. We reviewed the make_changes calls, and it seems to behave better with this change. 
We also agreed that it makes sense to refactor the common code in postgers_client

**Comment by romayalon** [comment]
Hi @aspandey,
As Danny said I was worried about some specific make_changes() calls that make multiple changes and the queries are not related to each other.
For example, something like - 
```
await system_store.make_changes({
        update: {
            buckets: buckets_updates,
            accounts: accounts_updates,
            pools: pools_updates,
            namespace_resources: namespace_resources_updates
        },
   });
```
I just wanted to make sure we won't roll back queries for no reason because this is the side effect of making one transaction per single make_changes() call instead of per collection.

File: src/test/unit_tests/test_system_store.js [hunk 7]
```diff
@@ -138,4 +138,37 @@ mocha.describe('system_store', function() {
...
+                    _id: system_id,
+                    name: orig_name,
+                    owner: system_store.new_system_store_id(),
+                }]
+            }
+        });
+        try {
+            await system_store.make_changes({
+                insert: {
+                    systems: [{
```
Comment (dannyzaken) [inline src/test/unit_tests/test_system_store.js:158]: For the second `make_changes` you should add a change\insert to different tables (pools, accounts, etc.). After the failure, make sure all changes were not applied. 
of course, this is the case for Postgres only, so we should skip this step for Mongo

File: src/server/system_services/system_store.js [hunk 8]
```diff
@@ -757,6 +756,7 @@ class SystemStore extends EventEmitter {
             }
         }
 
+
```
Comment (dannyzaken) [inline src/server/system_services/system_store.js:759]: Unnecessary change.
